fix(ivy): support SkipSelf when injecting ViewContainerRef, ElementRef and ChangeDetectorRef

crisbeto opened the pull request:
Currently Ivy doesn't support adding `@SkipSelf` for some of the special DI tokens (e.g. `ViewContainerRef`, `ElementRef`, `ChangeDetectorRef`). These changes add support for it since it's something we support in ViewEngine.

Fixes #34066.

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -334,7 +334,15 @@ export function getOrCreateInjectable<T>(
-      enterDI(lView, tNode);
+      if (flags & InjectFlags.SkipSelf) {
+        return getOrCreateInjectable(
+            tNode.parent, getParentInjectorView(getParentInjectorLocation(tNode, lView), lView),
```
@crisbeto I worked on a fix to support @SkipSelf for Injectors and noticed that `tNode.parent` may not be available when we cross component boundary, [see a test here](https://github.com/angular/angular/pull/34839/files#diff-881d964a81b3db8c8cb5311b8e3546ecR644). It looks like this code should be updated to handle this case as well ([similar to this line](https://github.com/angular/angular/pull/34839/files#diff-a29d4642602b40bdabe81990eed99e50R353)).)

crisbeto replied:
Hmm, would that be a problem for `ViewContainerRef`, `ChangeDetectorRef` and `ElementRef` though (crossing component boundaries)? A node should always be inside another element or a view container so we'd always resolve the value before we have to cross a component boundary. That being said, I might be missing some context here.

kara replied:
I believe @AndrewKushnir is talking about the case where the `child` element requesting `SkipSelf` is a `ViewChild` of `parent` instead of a `ContentChild`. 

```ts
@Component({selector: 'parent', template: '<child></child>' }) 
export class Parent {...}
```

If `child` is a root node of the `parent` template, `tNode.parent` would be null because the `parent` element is in a different view and we can't cross component boundaries with the `tNode` tree (because `tNode.parent` would change with each usage of the component).  So `ViewContainerRef` would not be found.

In this case, we should be falling back to the `T_HOST` saved in the `LView`:

```suggestion
            tNode.parent || lView[T_HOST}, getParentInjectorView(getParentInjectorLocation(tNode, lView), lView),
```

kara replied:
One more thing to consider: when crossing component boundaries, we have to respect `@Host` config flags as well. In other words, if there is `@SkipSelf` and `Host` together, the `ViewContainerRef` should not be found if it would cross a component boundary.

kara replied:
One more thing to consider: when crossing component boundaries, we have to respect `@Host` config flags as well. In other words, if there are `@SkipSelf` and `@Host` flags used together, the `ViewContainerRef` should not be found if it would cross a component boundary.

crisbeto committed:
fix(ivy): support SkipSelf when injecting ViewContainerRef, ElementRef and ChangeDetectorRef

Currently Ivy doesn't support adding `@SkipSelf` for some of the special DI tokens (e.g. `ViewContainerRef`, `ElementRef`, `ChangeDetectorRef`). These changes add support for it since it's something we support in ViewEngine.

Fixes #34066.

IgorMinar commented:
is this PR still relevant now that Andrew has #34839 in the merge queue?

crisbeto commented:
I think so. #34839 handles `SkipSelf` only for the `Injector` token, whereas these changes do it for `ElementRef`, `ViewContainerRef` etc.

crisbeto commented:
Rebased which removed the need to bump the size limit.

kara commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -698,6 +698,85 @@ describe('di', () => {
         expect(dirA.dirB.value).toEqual('parent');
       });
 
+      it('should support @SkipSelf when injecting ViewContainerRef', () => {
```
Can we add a test where the child element requesting `ViewContainerRef` is a ViewChild of the parent element?

kara reviewed (changes requested):
Sorry for the delayed review! Comments below

kara commented on packages/core/src/render3/di.ts:
```diff
@@ -333,7 +333,15 @@ export function getOrCreateInjectable<T>(
-      enterDI(lView, tNode);
+      if (flags & InjectFlags.SkipSelf) {
+        return getOrCreateInjectable(
+            tNode.parent, getParentInjectorView(getParentInjectorLocation(tNode, lView), lView),
```
One more thing to consider: when crossing component boundaries, we have to respect `@Host` config flags as well. In other words, if there are `@SkipSelf` and `@Host` flags used together, the `ViewContainerRef` should not be found if it would cross a component boundary.

AndrewKushnir commented:
FYI, it turned out that adding support or @Host decorator required extracting common logic into separate functions, so the commit from this PR is now cherry-pick'ed into #34839 and further review (and fixes if needed) will happen in #34839. Thank you.
